### PR TITLE
ARM64 build

### PR DESCRIPTION
**Pull request by alceasan**
ARM64 build for Raspberry Pi 4 (maybe can also work for 3B model, not tested). Bump ffmpeg version to 4.3 for stable hw accelerated encoding with h264_v4l2m2m codec.

### CODE REVIEW

File: docker/arm64/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,86 @@
+FROM alceasan/opencv-arm64:latest as opencv
```
Comment (alceasan) [inline docker/arm64/Dockerfile:1]: Please, change it to rolfcoopter repo, sorry

File: docker/arm64/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+FROM alceasan/ffmpeg-arm64:latest as ffmpeg
```
Comment (alceasan) [inline docker/arm64/Dockerfile:2]: Please, change it to rolfcoopter repo, sorry

**Comment by roflcoopter** [comment]
Looks good! One issue tho, i wont be able to build and push this locally since i dont have an RPi4 to do this... Not sure how to get around that. Any idea?

I have been working on a cross-build solution which would solve this problem tho, which i will release with the next version probably.

**Comment by alceasan** [comment]
My first build was using an ARM64 VM with ubuntu (https://ubuntu.com/download/server/arm) on my Desktop PC (AMD64 Linux). The performance was horrible, but it worked.

The best option IMHO would be using a multiarch build as you said, you can try using this: https://github.com/docker/build-push-action.

Meantime, I'm probably maintaining my fork with your changes because it's working well on a RPi4 ARM64.

**Comment by roflcoopter** [comment]
Sorry this is taking so long for me to implement, i cant find the time right now to finish this.
Thanks for maintaining your fork in the meantime

**Comment by roflcoopter** [comment]
@alceasan I finally got my multiarch builds to work, i put up a test-version under ```roflcoopter/viseron-new:docker_revamp```

Could you try and see if it works on your RPi4? Currently the generated ffmpeg command wont recognize its an RPi4 so the hwaccel args wont be added automatically

**Comment by alceasan** [comment]
I just tried it, but it's not working. Some problem with missing libraries it seems (libatomic.so.1), here is the log:

```
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.

[s6-init] ensuring user provided files have correct perms...exited 0.

[fix-attrs.d] applying ownership & permissions fixes...

[fix-attrs.d] done.

[cont-init.d] executing container initialization scripts...

[cont-init.d] 10-set-env-vars: executing...

****** Checking for hardware acceleration platforms ******

OpenCL cannot be used

VA-API cannot be used

CUDA cannot be used

*********************** Done *****************************

[cont-init.d] 10-set-env-vars: exited 0.

[cont-init.d] 20-check-if-rpi: executing...

********** Checking if we are running on an RPi **********

Running on an RPi4

*********************** Done *****************************

[cont-init.d] 20-check-if-rpi: exited 0.

[cont-init.d] done.

[services.d] starting services

[services.d] done.

Traceback (most recent call last):

  File "/src/viseron/viseron.py", line 6, in <module>

    from const import LOG_LEVELS

  File "/src/viseron/const.py", line 1, in <module>

    import cv2

  File "/usr/local/lib/python3.7/dist-packages/cv2/__init__.py", line 96, in <module>

    bootstrap()

  File "/usr/local/lib/python3.7/dist-packages/cv2/__init__.py", line 86, in bootstrap

    import cv2

ImportError: libatomic.so.1: cannot open shared object file: No such file or directory

[cont-finish.d] executing container finish scripts...

[cont-finish.d] done.

[s6-finish] waiting for services.

[s6-finish] sending all processes the TERM signal.

[s6-finish] sending all processes the KILL signal and exiting.
```

**Comment by roflcoopter** [comment]
> ImportError: libatomic.so.1: cannot open shared object file: No such file or directory

Fixed and pushed a new version of the same tag, Could you try again?

**Comment by alceasan** [comment]
Sorry about the delay, still not working, this time a problem with EdgeTPU libs:

```
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 10-adduser: executing... 
************************ UID/GID *************************
User uid:    1000
User gid:    1000
************************** Done **************************
[cont-init.d] 10-adduser: exited 0.
[cont-init.d] 20-set-env-vars: executing... 
****** Checking for hardware acceleration platforms ******
OpenCL cannot be used
VA-API cannot be used
CUDA cannot be used
*********************** Done *****************************
[cont-init.d] 20-set-env-vars: exited 0.
[cont-init.d] 30-check-if-rpi: executing... 
********** Checking if we are running on an RPi **********
Running on an RPi4
*********************** Done *****************************
[cont-init.d] 30-check-if-rpi: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
[2021-04-05 09:12:27] [root                    ] [INFO    ] - -------------------------------------------
[2021-04-05 09:12:27] [root                    ] [INFO    ] - Initializing...
[2021-04-05 09:12:28] [lib.mqtt                ] [INFO    ] - Initializing MQTT connection
[2021-04-05 09:12:28] [lib.detectors.edgetpu   ] [ERROR   ] - EdgeTPU not found. Detection will run on CPU
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/tflite_runtime/interpreter.py", line 152, in load_delegate
    delegate = Delegate(library, options)
  File "/usr/local/lib/python3.7/dist-packages/tflite_runtime/interpreter.py", line 111, in __init__
    raise ValueError(capture.message)
ValueError
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/src/viseron/lib/detectors/edgetpu/__init__.py", line 28, in __init__
    tflite.load_delegate("libedgetpu.so.1", {"device": "usb"})
  File "/usr/local/lib/python3.7/dist-packages/tflite_runtime/interpreter.py", line 155, in load_delegate
    library, str(e)))
ValueError: Failed to load delegate from libedgetpu.so.1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/tflite_runtime/interpreter.py", line 152, in load_delegate
    delegate = Delegate(library, options)
  File "/usr/local/lib/python3.7/dist-packages/tflite_runtime/interpreter.py", line 111, in __init__
    raise ValueError(capture.message)
ValueError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/src/viseron/lib/detectors/edgetpu/__init__.py", line 37, in __init__
    tflite.load_delegate("libedgetpu.so.1", {"device": "pci:0"})
  File "/usr/local/lib/python3.7/dist-packages/tflite_runtime/interpreter.py", line 155, in load_delegate
    library, str(e)))
ValueError: Failed to load delegate from libedgetpu.so.1
```

**Comment by roflcoopter** [comment]
> Sorry about the delay, still not working, this time a problem with EdgeTPU libs:
> 


Dont worry about it! I am the one whos late... ;)

Could you exec into the container and run some commands for me and then restart the container?
```
echo "deb https://packages.cloud.google.com/apt coral-edgetpu-stable main" | sudo tee /etc/apt/sources.list.d/coral-edgetpu.list
apt-get update
apt-get --reinstall install libedgetpu1-max=15.0
```

There were major version problems going on when google switched the tflite_runtime to a new project name, im guessing this might be an issue

**Comment by alceasan** [comment]
Sorry, but after installing the package and restarting container, it's showing the same error :(

```
# apt info libedgetpu1-max
Package: libedgetpu1-max
Version: 15.0
Priority: optional
Section: misc
Source: libedgetpu
Maintainer: Coral <coral-support@google.com>
Installed-Size: 882 kB
Provides: libedgetpu1 (= 15.0)
Depends: libc6, libgcc1, libstdc++6, libusb-1.0-0, debconf (>= 0.5) | debconf-2.0
Conflicts: libedgetpu1, libedgetpu1-legacy
Homepage: https://coral.ai/
Download-Size: 259 kB
APT-Manual-Installed: yes
APT-Sources: https://packages.cloud.google.com/apt coral-edgetpu-stable/main arm64 Packages
Description: Support library for Edge TPU
 Support library (max speed) for the Edge TPU

```

**Comment by roflcoopter** [comment]
Actually scratch my last comment, it wont solve anything.
The error printed is just the caught stacktrace when no EdgeTPU is found. Did you start the container with ```-v /dev/bus/usb:/dev/bus/usb --privileged```?

**Comment by alceasan** [comment]
The container is already running with that configuration, but I'm not using any Coral USB.

**Comment by roflcoopter** [comment]
> The container is already running with that configuration, but I'm not using any Coral USB.

Aha i see! Well that means it should work then, because the stacktrace you linked is printed by [this line](https://github.com/roflcoopter/viseron/blob/66afba54b112b8bfabfdf6e2a6a2999d5efa2484/src/lib/detectors/edgetpu/__init__.py#L42-L44). Viseron is not actually crashing at that point.

**Comment by alceasan** [comment]
You are right, even after that errors container is running. I have tested it a little bit, but it haven't registered any movement on MQTT or recordings, so it's not really working.

Also, it has a much higher cpu usage compared with my working version on arm64 as you can see in the netdata monitoring after changing viseron container from one image to other:

image

This is probably because of the ffmpeg codec used. I'm using a ffmpeg build with the v4l2 codec, which has hardware acceleration on RPI4 (arm64), using this configuration: https://www.willusher.io/general/2020/11/15/hw-accel-encoding-rpi4

**Comment by roflcoopter** [comment]
The container is built with `v4l2` codec, as you can see [here](https://github.com/roflcoopter/viseron/blob/99ad9ae4ec7f7838ecf0d24240179549a0a9f455/docker/aarch64/Dockerfile.ffmpeg#L497). 

However, i found a typo in one of the environment variables so the codec was not automatically picked up. Could you pull the new `roflcoopter/viseron:dev` and show me the complete log output?

**Comment by alceasan** [comment]
It's not working with that image, but it seems it's due to a config error. has something changed in the config for this version?

```
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.

[s6-init] ensuring user provided files have correct perms...exited 0.

[fix-attrs.d] applying ownership & permissions fixes...

[fix-attrs.d] done.

[cont-init.d] executing container initialization scripts...

[cont-init.d] 10-adduser: executing... 

usermod: no changes

************************ UID/GID *************************

User uid:    1000

User gid:    1000

************************** Done **************************

[cont-init.d] 10-adduser: exited 0.

[cont-init.d] 20-set-env-vars: executing... 

****** Checking for hardware acceleration platforms ******

OpenCL cannot be used

VA-API cannot be used

CUDA cannot be used

*********************** Done *****************************

[cont-init.d] 20-set-env-vars: exited 0.

[cont-init.d] 30-check-if-rpi: executing... 

********** Checking if we are running on an RPi **********

Running on an RPi4

*********************** Done *****************************

[cont-init.d] 30-check-if-rpi: exited 0.

[cont-init.d] done.

[services.d] starting services

[services.d] done.

Traceback (most recent call last):

  File "/src/viseron/viseron.py", line 8, in <module>

    from lib.config import CONFIG, NVRConfig, ViseronConfig

  File "/src/viseron/lib/config/__init__.py", line 94, in <module>

    CONFIG = VISERON_CONFIG_SCHEMA(raw_config)

  File "/usr/local/lib/python3.7/dist-packages/voluptuous/schema_builder.py", line 272, in __call__

    return self._compiled([], data)

  File "/usr/local/lib/python3.7/dist-packages/voluptuous/schema_builder.py", line 594, in validate_dict

    return base_validate(path, iteritems(data), out)

  File "/usr/local/lib/python3.7/dist-packages/voluptuous/schema_builder.py", line 432, in validate_mapping

    raise er.MultipleInvalid(errors)

voluptuous.error.MultipleInvalid: required key not provided @ data['cameras'][0]['substream']['port']

[cont-finish.d] executing container finish scripts...

[cont-finish.d] done.

[s6-finish] waiting for services.

[s6-finish] sending all processes the TERM signal.

[s6-finish] sending all processes the KILL signal and exiting.
```

**Comment by roflcoopter** [comment]
Yes, a `port` has to be specified if you are using substream

**Comment by roflcoopter** [comment]
> OK, I included the port also in the substream config, but it seems there is a problem with ffmpeg. I have two cameras in my config, here is the current log after starting (it's repeating every few seconds):
> 
> ```
> [2021-04-06 16:07:09] [lib.camera.camera1     ] [ERROR   ] - Restarting frame pipe
> [2021-04-06 16:07:11] [lib.camera.camera2   ] [ERROR   ] - Unable to decode frame. FFmpeg pipe seems broken
> [2021-04-06 16:07:11] [lib.camera.camera2   ] [ERROR   ] - Unable to decode frame. FFmpeg pipe seems broken, message repeated 2 times
> [2021-04-06 16:07:11] [lib.camera.camera2   ] [ERROR   ] - Unable to decode frame. FFmpeg pipe seems broken, message repeated 3 times
> [2021-04-06 16:07:11] [lib.camera.camera2   ] [ERROR   ] - Unable to decode frame. FFmpeg pipe seems broken, message repeated 4 times
> [2021-04-06 16:07:11] [lib.camera.camera2   ] [ERROR   ] - Unable to decode frame. FFmpeg pipe seems broken, message repeated 5 times
> [2021-04-06 16:07:12] [lib.camera.camera1      ] [ERROR   ] - Unable to decode frame. FFmpeg pipe seems broken
> [2021-04-06 16:07:12] [lib.camera.camera1      ] [ERROR   ] - Unable to decode frame. FFmpeg pipe seems broken, message repeated 2 times
> [2021-04-06 16:07:12] [lib.camera.camera1      ] [ERROR   ] - Unable to decode frame. FFmpeg pipe seems broken, message repeated 3 times
> [2021-04-06 16:07:12] [lib.camera.camera1      ] [ERROR   ] - Unable to decode frame. FFmpeg pipe seems broken, message repeated 4 times
> [2021-04-06 16:07:12] [lib.camera.camera1      ] [ERROR   ] - Unable to decode frame. FFmpeg pipe seems broken, message repeated 5 times
> ```
> 
> How could I get a more detailed log?

A good way to start is to set `ffmpeg_loglevel: info`, maybe turn on debug logging for Viseron aswell.

**Comment by roflcoopter** [comment]
Closing as RPi4 support has been implemented some time ago

**Pull request closed without merging**